source/common/http/utility.cc

```diff
@@ -148,39 +149,18 @@ bool Utility::isWebSocketUpgradeRequest(const HeaderMap& headers) {
                     Http::Headers::get().UpgradeValues.WebSocket.c_str())));
 }
 
-Http2Settings Utility::parseHttp2Settings(const Json::Object& config) {
+Http2Settings Utility::parseHttp2Settings(const envoy::api::v2::Http2ProtocolOptions& config) {
   Http2Settings ret;
-
-  Json::ObjectSharedPtr http2_settings = config.getObject("http2_settings", true);
-  ret.hpack_table_size_ =
-      http2_settings->getInteger("hpack_table_size", Http::Http2Settings::DEFAULT_HPACK_TABLE_SIZE);
-  ret.max_concurrent_streams_ = http2_settings->getInteger(
-      "max_concurrent_streams", Http::Http2Settings::DEFAULT_MAX_CONCURRENT_STREAMS);
-  ret.initial_stream_window_size_ = http2_settings->getInteger(
-      "initial_stream_window_size", Http::Http2Settings::DEFAULT_INITIAL_STREAM_WINDOW_SIZE);
+  ret.hpack_table_size_ = MessageUtil::getUInt32(config, "hpack_table_size",
```

Comment:
This came about as a result of wanting to reduce the boiler plate around returning default values if a field is missing (or throw an exception). Given we have O(100s) of these situations, it made sense to try and factor out this boiler plate.

There's two ways to do this:
1. Via reflection, as above, where we do dynamic stuff to figure out whether a field is set and return the real/default value or exception.
2. Via C preprocessor macros. We could generate the code for `if (config.has_hpack_tablle_size()) { return config.hpack_table_size(); } else { return default_value; }`.

@lizan has expressed preference for (2) since its faster (which isn't an issue in this particular code) and more type safe (which is a win). I'm not a huge fan of macros in modern C++. I can go either way though, since you do get a type safety win by using them here.